pkg/operator/controller/ingress/status.go

```diff
@@ -644,21 +645,41 @@ func formatConditions(conditions []*operatorv1.OperatorCondition) string {
 // computeIngressProgressingCondition computes the ingresscontroller's
 // "Progressing" status condition, which indicates if the ingresscontroller's
 // current state matches its desired state.
-func computeIngressProgressingCondition(conditions []operatorv1.OperatorCondition, ic *operatorv1.IngressController, service *corev1.Service, platform *configv1.PlatformStatus) operatorv1.OperatorCondition {
+func computeIngressProgressingCondition(conditions []operatorv1.OperatorCondition, ic *operatorv1.IngressController, service *corev1.Service, platform *configv1.PlatformStatus, deployment *appsv1.Deployment) operatorv1.OperatorCondition {
```

Comment:
I wonder if it's necessary to recompute the replicas when we already have conditions for `computeDeploymentReplicasMinAvailableCondition` and `computeDeploymentReplicasAllAvailableCondition`.  Perhaps you could just add one more method `computeDeploymentReplicasAllUpdatedCondition` and use those conditions to determine Progressing status?

Reply:
Yea I'll admit this is pretty confusing to me, I've gone back and forth. cluster-ingress-operator and cluster-dns-operator do things in different patterns. But you've hit on a good point.

So I think those functions are doing slightly different things. It looks like they always add individual `status.condition` entries and are used for calculating the "degraded" (`operatorv1.OperatorStatusTypeDegraded`) in `oc get co` clusteroperator status. So each function adds independent `status.condition`.

Right now, this new logic stuffs all of the "progressing" errors messages in to 1 single condition, the progressing condition (`operatorv1.OperatorStatusTypeProgressing`). We could make a new `computeDeploymentProgressingCondition` function which would make a new individual `status.condition` (e.g. `DeploymentProgressing`) that would always need to be there. Then we could calculate `operatorv1.OperatorStatusTypeProgressing` just like `operatorv1.OperatorStatusTypeDegraded` is calculated.

However, it would involve a little bit of refactoring. Since the previous developers just packed (lazily? or intentionally?) the `operatorv1.OperatorStatusTypeProgressing` condition with `LoadBalancerStatusIncomplete` and `NoService`, these would need to become new individual status conditions (e.g. stop overloading `operatorv1.OperatorStatusTypeProgressing`).

I'm conflicted. I haven't found any standards on error messages. For example, `cluster-dns-operator` doesn't make individual condition entries for each status. It just packs the progressing, degraded, available condition messages with multiple appended error messages (https://github.com/openshift/cluster-dns-operator/blob/master/pkg/operator/controller/dns_status.go#L66-L70). 

I think the important thing here, is that the "real" status messages are `OperatorStatusTypeAvailable`, `OperatorStatusTypeProgressing`, and `OperatorStatusTypeDegraded` because these types of conditions determine the status of `oc get co`. I suppose we can do it however we want. If we want multiple individual error conditions, then aggregate it into a single `OperatorStatusType*` condition, I suppose that is fine.

Sorry for the long response....I had to spend a lot of time working this out in my head. I think I'm indifferent. I lean towards making them individual conditions because that's how `OperatorStatusTypeAvailable` and `OperatorStatusTypeProgressing` are done.

Reply:
Not sure if I'm following. I don't think the existing `status.conditions` offer a condition based on `status.UpdatedReplicas`. For reference, the `kubectl rollout` command uses mostly `status.UpdatedReplicas` as well as `status.AvailableReplicas` to determine if pod is actively rolling out:
https://github.com/kubernetes/kubectl/blob/master/pkg/polymorphichelpers/rollout_status.go#L76-L89. I think that's what we should use here to determine the progressing status.

Both `computeDeploymentReplicasMinAvailableCondition` and `computeDeploymentReplicasAllUpdatedCondition` compute a status solely based on `Status.AvailableReplicas` as well as the `deployment.Spec.Strategy` specifiers, and though closely related, I don't think that's completely correct when calculating the progressing condition. I don't think either of these functions/conditions tell us if a pod is progressing.

Specifically `computeDeploymentReplicasAllUpdatedCondition` is very close to telling us if the deployment is progressing, but it just checks `deployment.Status.AvailableReplicas < replicas` which isn't enough info. You could be still rolling out and not trigger this condition. `computeDeploymentReplicasMinAvailableCondition` just says if we don't have enough replicas running, doesn't tell us if it's progressing.

Reply:
> Perhaps you could just add one more method `computeDeploymentReplicasAllUpdatedCondition` and use those conditions to determine Progressing status?

Reply:
What do you mean by "those conditions"? Are you suggesting to add a new `DeploymentReplicasAllUpdated` condition `status.condition.type` and use that new condition to calculate the progressing status?

Or are you suggesting to use the existing `status.conditions` generated by `computeDeploymentReplicasMinAvailableCondition` and `computeDeploymentReplicasAllUpdatedCondition` to calculate the progressing status?

Reply:
Add a new method called `computeDeploymentReplicasAllUpdatedCondition` which does what you do in:
https://github.com/openshift/cluster-ingress-operator/blob/ec2c9bcada6a46026cc3f5e758b7c03565dc056f/pkg/operator/controller/ingress/status.go#L667-L671
Call it before you call `computeIngressProgressingCondition`, maybe after L76.  Then, instead of passing the `deployment` variable into `computeIngressProgressingCondition`, you can look up the `conditions` to get the information you need to mark the `DeploymentProgressing` condition.